src/ws/ws_manager.rs

```diff
@@ -95,21 +109,26 @@ impl WsManager {
         let subscriptions = Arc::new(Mutex::new(subscriptions_map));
         let subscriptions_copy = Arc::clone(&subscriptions);
 
-        let reader_fut = async move {
-            // TODO: reconnect
-            loop {
-                let data = reader.next().await;
-                if let Err(err) = WsManager::parse_and_send_data(data, &subscriptions_copy).await {
-                    error!("Error processing data received by WS manager reader: {err}");
+        let reader_handle = {
+            let stop_flag1 = Arc::clone(&stop_flag);
+            let reader_fut = async move {
+                // TODO: reconnect
+                while !stop_flag1.load(Ordering::Relaxed) {
+                    let data = reader.next().await;
+                    if let Err(err) = WsManager::parse_and_send_data(data, &subscriptions_copy).await {
+                        error!("Error processing data received by WS manager reader: {err}");
+                    }
                 }
-            }
+                warn!("ws message reader task stopped");
+            };
+            spawn(reader_fut)
         };
-        spawn(reader_fut);
 
-        {
+        let ping_handle = {
+            let stop_flag2 = Arc::clone(&stop_flag);
```

Comment:
do `let stop_flag = Arc::clone(&stop_flag)` here and above. the original stop_flag will be shadowed

Reply:
Oh you meant shadowing. Got it.